### PR TITLE
feat: support pydantic v2 [APE-1372]

### PR DESCRIPTION
**Pull request by banteg**
### What I did

support pydantic 2 in a backwards-compatible way.

### How I did it

try to import from `pydantic.v1` (works in v2) and fall back to `pydantic` (works in v1)

relax the dependency so it matches both v1 and v2.

this is a quick solution that allows us to keep compat with other packages that are starting to update to pydantic v2 already.

the full migration will require a more involved refactor.

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -74,7 +74,7 @@
     include_package_data=True,
     install_requires=[
         "hexbytes>=0.3.0,<1",
-        "pydantic>=1.10.7,<2",
+        "pydantic>=1.10.7",  # pydantic >=2 is supported
```
Comment (fubuloubu) [inline setup.py:77]: Maybe cap on `<3`

**Review by fubuloubu — APPROVED** [review]

File: setup.py (hunk 1 above)
Comment (banteg) [reply setup.py:77]: im not fan of upper caps, they always cause problems

Comment (fubuloubu) [reply setup.py]: When they upgrade to v3, they will drop the v1 backports

File: setup.py [hunk 2]
```diff
@@ -74,7 +74,7 @@
     include_package_data=True,
     install_requires=[
         "hexbytes>=0.3.0,<1",
-        "pydantic>=1.10.7,<2",
+        "pydantic>=1.10.7,<3",
```
Comment (antazoey) [inline setup.py:77]: we may want to make 2.3 the smallest 2- version we support (exclude 2.0-2.2)

**Review by fubuloubu — DISMISSED** [review]

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**